File: C++/代码随想录/Greedy/Dimensional_tradeoff/135. Candy.cpp
// 135. Candy
// https://leetcode.com/problems/candy/

#include <iostream>
#include <vector>
#include <numeric>

int candy(std::vector<int>& ratings) {
    int n = ratings.size();
    std::vector<int> candies(n, 1);
    
    for (int i = 1; i < n; i++) {
        if (ratings[i] > ratings[i - 1]) {
            candies[i] = candies[i - 1] + 1;
        }
    }
    
    for (int i = n - 2; i >= 0; i--) {
        if (ratings[i] > ratings[i + 1]) {
            candies[i] = std::max(candies[i], candies[i + 1] + 1);
        }
    }

    return std::accumulate(candies.begin(), candies.end(), 0);
}

int main() {
    std::vector<int> ratings = {1, 0, 2}; // Expected output:5
    std::cout << "Minimum number of candies: " << candy(ratings) << std::endl;

    std::vector<int> ratings2 = {1, 2, 2}; // Expected output:4
    std::cout << "Minimum number of candies: " << candy(ratings2) << std::endl;
    
    return 0;
}

File: C++/代码随想录/Greedy/Dimensional_tradeoff/135. Candy.py
def candy(ratings):
    n = len(ratings)
    candies = [1] * n

    # 从左到右遍历，确保每个孩子比他左边评分低的孩子得到更多的糖果
    for i in range(1, n):
        if ratings[i] > ratings[i - 1]:
            candies[i] = candies[i - 1] + 1

    # 从右到左遍历，确保每个孩子比他右边评分低的孩子得到更多的糖果
    for i in range(n - 2, -1, -1):
        if ratings[i] > ratings[i + 1]:
            candies[i] = max(candies[i], candies[i + 1] + 1)

    return sum(candies)

# 测试用例
print(candy([1, 0, 2]))  # 应该输出 5
print(candy([1, 2, 2]))  # 应该输出 4
